File: executor/executor/images.py
from executor.utils import log
from pathlib import Path
from zstandard import ZstdDecompressor
import hashlib
import requests
import shutil
import tempfile
import threading
import time
import typing


IMAGES_SERVER_POLL_INTERVAL = 5 * 60  # 5 minutes


class ImagesRetriever:
    def __init__(self, cli):
        self._http = requests.Session()
        self._server = cli.images_server.rstrip("/")

        if cli.images_cache_dir is not None:
            self._storage_dir: Path = cli.images_cache_dir
            self._storage_dir.mkdir(parents=True, exist_ok=True)
        else:
            # If no cache dir is configured, create a temporary one just for this invocation. This
            # avoids having separate code paths for "cached" and "not cached".
            self._storage_dir = Path(tempfile.mkdtemp())

        self._latest_commit = self._get_text("latest")
        self._purge_old_caches()

    def get_image(self, name):
        local_path = self._storage_dir / self._latest_commit / f"{name}.qcow2"
        local_path.parent.mkdir(exist_ok=True, parents=True)

        image_url = f"images/{self._latest_commit}/{name}.qcow2"

        if not local_path.exists():
            log(f"downloading image {name} (commit: {self._latest_commit})")

            resp = self._http.get(f"{self._server}/{image_url}.zst", stream=True)
            resp.raise_for_status()

            decompressor = ZstdDecompressor()
            with local_path.open("wb") as dst:
                src = typing.cast(typing.BinaryIO, resp.raw)
                decompressor.copy_stream(src, dst)

        # Check that the image we are running matches the hash the images server expect. This helps
        # detect tampering in the images cache (possibly done by a compromised previous build).
        log(f"verifying hash of image {name}")
        local_hash = hashlib.file_digest(local_path.open("rb"), "sha256").hexdigest()
        remote_hash = self._get_text(f"{image_url}.sha256")
        if local_hash != remote_hash:
            print(f"error: local hash of image {name} differs from the remote one")
            print(f"local hash: {local_hash}")
            print(f"remote hash: {remote_hash}")
            exit(1)

        return local_path

    def _get_text(self, path):
        resp = self._http.get(f"{self._server}/{path}")
        resp.raise_for_status()
        return resp.text.strip()

    def _purge_old_caches(self):
        for entry in self._storage_dir.iterdir():
            if entry.is_dir() and entry.name != self._latest_commit:
                log(f"purging image cache for commit {entry.name}")
                shutil.rmtree(entry)


class ImageUpdateWatcher(threading.Thread):
    def __init__(self, retriever: ImagesRetriever, then):
        self._retriever = retriever
        self._then = then

        super().__init__(name="image-update-watcher", daemon=True)

    def run(self):
        log("started polling the image server to check for image updates")
        while True:
            time.sleep(IMAGES_SERVER_POLL_INTERVAL)
            try:
                new_commit = self._retriever._get_text("latest")
            except requests.exceptions.RequestException as e:
                print(f"warn: failed to check for image updates: {e}")
                continue
            if new_commit != self._retriever._latest_commit:
                log(f"new images with commit {new_commit} are available")
                self._then()

File: executor/src/images.rs
use crate::qemu::{ShutdownReason, ShutdownRequester};
use anyhow::{Error, anyhow};
use reqwest::blocking::Client;
use sha2::{Digest as _, Sha256};
use std::fs::{File, remove_dir_all};
use std::io::{BufReader, Read};
use std::path::{Path, PathBuf};
use std::sync::OnceLock;
use std::time::Duration;
use tempfile::TempDir;
use zstd::stream::read::Decoder;

pub struct ImagesRetriever {
    http: Client,
    server_url: String,
    storage_dir: PathBuf,
    latest_commit: OnceLock<String>,

    _tempdir_drop_guard: Option<TempDir>,
}

impl ImagesRetriever {
    pub fn new(server_url: &str, cache_dir: Option<&Path>) -> Result<Self, Error> {
        // If no cache dir is configured, create a temporary one just for this invocation. This
        // avoids having separate code paths for "cached" and "not cached".
        let mut tempdir_drop_guard = None;
        let storage_dir = if let Some(path) = cache_dir {
            std::fs::create_dir_all(path)?;
            path.to_path_buf()
        } else {
            let tempdir = TempDir::new()?;
            let tempdir_path = tempdir.path().to_path_buf();
            tempdir_drop_guard = Some(tempdir);
            tempdir_path
        };

        Ok(Self {
            http: Client::new(),
            server_url: server_url.trim_end_matches('/').to_string(),
            storage_dir,
            latest_commit: OnceLock::new(),
            _tempdir_drop_guard: tempdir_drop_guard,
        })
    }

    pub fn get_image(&self, name: &str) -> Result<PathBuf, Error> {
        let commit = self.latest_commit()?;

        let path = self.storage_dir.join(&commit).join(format!("{name}.qcow2"));
        std::fs::create_dir_all(path.parent().unwrap())?;

        let url = self.url(&format!("images/{commit}/{name}.qcow2.zst"));
        if !path.exists() {
            eprintln!("downloading image {name} (commit: {commit})");

            let resp = self.http.get(&url).send()?.error_for_status()?;
            std::io::copy(&mut Decoder::new(resp)?, &mut File::create(&path)?)?;
        }

        // Check that the image we are running matches the hash the images server expect. This helps
        // detect tampering in the images cache (possibly done by a compromised previous build).
        eprintln!("verifying hash of image {name}");
        let local_hash = sha256_file(&path)?;
        let remote_hash = self.retrieve_text(&format!("images/{commit}/{name}.qcow2.sha256"))?;
        if local_hash != remote_hash {
            return Err(anyhow!("local hash: {local_hash}")
                .context(format!("remote hash: {remote_hash}"))
                .context(format!(
                    "local hash of image {name} differs from the remote one"
                )));
        }

        Ok(path)
    }

    pub fn purge_old_caches(&self) -> Result<(), Error> {
        let latest_commit = self.latest_commit()?;
        for entry in self.storage_dir.read_dir()? {
            let path = entry?.path();
            if path.is_dir() && path.file_name().and_then(|s| s.to_str()) != Some(&latest_commit) {
                eprintln!(
                    "purging outdated image cache for commit {}",
                    path.file_name().unwrap().to_string_lossy()
                );
                remove_dir_all(path)?;
            }
        }
        Ok(())
    }

    fn latest_commit(&self) -> Result<&str, Error> {
        if let Some(value) = self.latest_commit.get() {
            return Ok(value);
        }
        let commit = self.retrieve_text("latest")?;
        Ok(self.latest_commit.get_or_init(move || commit))
    }

    fn retrieve_text(&self, url: &str) -> Result<String, Error> {
        Ok(self
            .http
            .get(self.url(url))
            .send()?
            .error_for_status()?
            .text()?
            .trim()
            .to_string())
    }

    fn url(&self, path: &str) -> String {
        format!("{}/{path}", self.server_url)
    }
}

pub fn watch_for_image_updates(
    retriever: ImagesRetriever,
    shutdown_requester: ShutdownRequester,
) -> Result<(), Error> {
    let latest_commit = retriever.latest_commit()?.to_string();

    eprintln!("started polling the image server to check for image updates");
    std::thread::spawn(move || {
        loop {
            std::thread::sleep(Duration::from_secs(60));
            match retriever.retrieve_text("latest") {
                Ok(new_commit) if new_commit == latest_commit => {}
                Ok(new_commit) => {
                    eprintln!("new images with commit {new_commit} are available");
                    shutdown_requester.request_shutdown(ShutdownReason::NewImages);
                }
                Err(err) => {
                    eprintln!("warning: failed to fetch the latest image commit: {err}");
                }
            }
        }
    });
    Ok(())
}

fn sha256_file(file: &Path) -> Result<String, Error> {
    let mut file = BufReader::new(File::open(file)?);
    let mut hasher = Sha256::new();

    let mut buf = vec![0; 1024 * 1024 * 4];
    loop {
        match file.read(&mut buf)? {
            0 => break,
            len => hasher.update(&buf[..len]),
        }
    }

    Ok(hex::encode(hasher.finalize().as_slice()))
}
